gensim/models/_utils_any2vec.pyx

```diff
@@ -92,3 +92,66 @@ cpdef compute_ngrams(word, unsigned int min_n, unsigned int max_n):
         for i in range(0, len(extended_word) - ngram_length + 1):
             ngrams.append(extended_word[i:i + ngram_length])
     return ngrams
+
+#
+# UTF-8 bytes that begin with 10 are subsequent bytes of a multi-byte sequence,
+# as opposed to a new character.
+#
+_MB_MASK = 0xC0
+_MB_START = 0x80
+
+
+cdef _is_utf8_continue_py3(b):
+    return b & _MB_MASK == _MB_START
+
+
+cdef _is_utf8_continue_py2(b):
+    return ord(b) & _MB_MASK == _MB_START
+
+
+_is_utf8_continue = _is_utf8_continue_py2 if PY2 else _is_utf8_continue_py3
+
+
+cpdef compute_ngrams_bytes(word, unsigned int min_n, unsigned int max_n):
```

Comment:
I check this current code on benchmark from @menshikh-iv  and find that it's slower than code with `split+join` (3.7s vs 5.6s). In C-code there are quite a lot of callings of python functions.

```suggestion
cdef unsigned char _MB_MASK = 0xC0
cdef unsigned char _MB_START = 0x80


cpdef compute_ngrams_bytes_my(word, unsigned int min_n, unsigned int max_n):
    """Computes ngrams for a word.
    Ported from the original FB implementation.
    Parameters
    ----------
    word : str
        A unicode string.
    min_n : unsigned int
        The minimum ngram length.
    max_n : unsigned int
        The maximum ngram length.
    Returns:
    --------
    list of bytes
        A list of ngrams, where each ngram is a list of **bytes**.
    See Also
    --------
    `Original implementation <https://github.com/facebookresearch/fastText/blob/7842495a4d64c7a3bb4339d45d6e64321d002ed8/src/dictionary.cc#L172>`__
    """
    cdef bytes utf8_word = f'<{word}>'.encode("utf-8")
    cdef const unsigned char* utf8_word_ptr = utf8_word
    cdef size_t num_bytes = len(utf8_word)
    cdef size_t j, i, n

    ngrams = []
    for i in range(num_bytes):
        if utf8_word_ptr[i] & _MB_MASK == _MB_START:
            continue

        j, n = i, 1
        while j < num_bytes and n <= max_n:
            j += 1
            while j < num_bytes and (utf8_word_ptr[i] & _MB_MASK) == _MB_START:
                j += 1
            if n >= min_n and not (n == 1 and (i == 0 or j == num_bytes)):
                ngrams.append(utf8_word[i:j])
            n += 1
    return ngrams
```

I don't check it on python2, but think that it should works fine. It's better to recheck (bemchark) my code, current code and classic variant (with split+join of str) again.

Reply:
There are two problems with the above code, both related to Py2:

- fstrings won't work in Python 2
- utf8_word_ptr[i] & _MB_MASK won't work in Python 2, because of str to int comparison. That's why you need to use byte_to_int function.

Reply:
The benchmarking results look good: 

```
(gensim) misha@cabron:~/git/gensim$ ipython -i bench.py 
Python 3.7.1 (default, Oct 22 2018, 11:21:55) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.2.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from gensim.models._utils_any2vec import compute_ngrams_bytes_optimized

In [2]: from gensim.models._utils_any2vec import compute_ngrams_bytes

In [3]: %time benchmark(ngram_func=compute_ngrams_bytes)
CPU times: user 956 ms, sys: 2.66 ms, total: 958 ms
Wall time: 957 ms

In [4]: %time benchmark(ngram_func=compute_ngrams_bytes_optimized)
CPU times: user 614 ms, sys: 103 µs, total: 614 ms
Wall time: 612 ms
```

Reply:
@horpto Thank you for helping with this!

Reply:
@mpenkov Ok, great!

But I wonder why `bytes(bytez[i:j])` is faster than original `utf8_word[i:j]` =(. `bytes(bytez[i:j])` allocates new memory and make memcpy unlike `utf8_word[i:j]`.
Original code generates `PySequence_GetSlice` call. Maybe it has a more overhead than I'd thouht. Cython has all the cards to generate a more concise call, not general PySequence.